File: app/src/main/java/info/jukov/vkstoryteller/MathUtils.java
package info.jukov.vkstoryteller;

import android.graphics.Rect;
import android.util.Log;
import android.util.Pair;
import java.util.Collection;
import java.util.StringTokenizer;

/**
 * User: jukov
 * Date: 06.09.2017
 * Time: 22:31
 */

public final class MathUtils {


	private static final String TAG = "MathUtils";

	public static float getDistanceBetweenTwoPoints(final float x1, final float y1, final float x2, final float y2) {
		final float catheter1 = Math.abs(x1 - x2);
		final float catheter2 = Math.abs(y1 - y2);

		return (float) Math.hypot(catheter1, catheter2);
	}

	public static float getScalarProduct(
		final float x1Source,
		final float y1Source,
		final float x1Dest,
		final float y1Dest,
		final float x2Source,
		final float y2Source,
		final float x2Dest,
		final float y2Dest) {

		final float x1 = x1Dest - x1Source;
		final float x2 = x2Dest - x2Source;
		final float y1 = y1Dest - y1Source;
		final float y2 = y2Dest - y2Source;
		return x1*x2 + y1*y2;
	}

	public static float getAverageAngleBetweenPointsPairsAndCentroid(final float[] beginPoints, final float[] endPoints) {

		final Pair<Float, Float> centroid = getCentroid(beginPoints);

		float angleSum = 0;

		for (int i = 0; i < beginPoints.length; i += 2) {
			final float beginPointToCentroidLength = getDistanceBetweenTwoPoints(
				beginPoints[i],
				beginPoints[i + 1],
				centroid.first,
				centroid.second
			);

			final float scalarProduct = getScalarProduct(
				centroid.first,
				centroid.second,
				beginPoints[i],
				beginPoints[i + 1],
				centroid.first,
				centroid.second,
				endPoints[i],
				endPoints[i + 1]
			);

			final float vector1Length = getDistanceBetweenTwoPoints(
				centroid.first,
				centroid.second,
				beginPoints[i],
				beginPoints[i + 1]
			);

			final float vector2Length = getDistanceBetweenTwoPoints(
				centroid.first,
				centroid.second,
				endPoints[i],
				endPoints[i + 1]
			);

			final float cos = scalarProduct / (vector1Length * vector2Length);
			final float angle = (float) Math.toDegrees(Math.acos(cos));

			final float signedAngle = angle * getVectorDirectionOnCircle(
				endPoints[i],
				endPoints[i + 1],
				beginPoints[i],
				beginPoints[i + 1],
				centroid
			);

			angleSum += angle;
		}

		return angleSum / (beginPoints.length / 2);
	}

	public static float getAverageDistanceFromPointsToCentroid(final float[] points) {
		if (points.length < 2) {
			throw new IllegalStateException("Require at least 1 point for caclulate distance");
		}

		final Pair<Float, Float> centroid = getCentroid(points);

		float distance = 0;

		for (int i = 0; i < points.length; i += 2) {
			distance += getDistanceBetweenTwoPoints(
				points[i],
				points[i + 1],
				centroid.first,
				centroid.second
			);
		}

		return distance / (points.length / 2);
	}

	public static boolean isPointerInBounds(final Pair<Float, Float> pointerCoordinates, final Rect bounds) {
		return bounds.contains(pointerCoordinates.first.intValue(), pointerCoordinates.second.intValue());
	}

	public static Pair<Float, Float> getCentroid(final float[] points) {

		float centroidX = 0;
		float centroidY = 0;

		for (int i = 0; i < points.length; i += 2) {
			centroidX += points[i];
			centroidY += points[i + 1];
		}

		final int pointsCount = points.length / 2;
		return new Pair<Float, Float>(centroidX / pointsCount, centroidY / pointsCount);
	}

	public static Pair<Float, Float> getCentroid(final Collection<Pair<Float, Float>> points) {

		float centroidX = 0;
		float centroidY = 0;

		for (final Pair<Float, Float> point : points) {
			centroidX += point.first;
			centroidY += point.second;
		}

		return new Pair<Float, Float>(centroidX / points.size(), centroidY / points.size());
	}

	private static float getVectorDirectionOnCircle(
		final float xSource,
		final float ySource,
		final float xDest,
		final float yDest,
		final Pair<Float, Float> circleCenter) {

		if (circleCenter.second > yDest) {

		} else {

		}

	}

	private MathUtils() {
	}
}

File: app/src/main/java/info/jukov/vkstoryteller/surface/MathUtils.kt
package info.jukov.vkstoryteller.surface

/**
 * User: jukov
 * Date: 06.09.2017
 * Time: 22:31
 *
 * Методы работют с массивами точек, координаты которых рассположены последовательно. Например:
 * Points [x1, y1, x2, y2, ... xN, yN]
 * Vector [xA, yA, xB, yB]
 */

fun getDistanceBetweenTwoPoints(point1: FloatArray, point2: FloatArray): Float {
    require(point1.size == 2 && point2.size == 2, { "Arrays must be contain only one point" })

    val catheter1 = Math.abs(point1[0] - point2[0])
    val catheter2 = Math.abs(point1[1] - point2[1])

    return Math.hypot(catheter1.toDouble(), catheter2.toDouble()).toFloat()
}

fun getScalarProduct(vector1: FloatArray, vector2: FloatArray): Float {
    require(vector1.size == 4 && vector2.size == 4, { "Arrays must be contain only two points of begining and ending of vector" })

    val x1 = vector1[2] - vector1[0]
    val x2 = vector2[2] - vector2[0]
    val y1 = vector1[3] - vector1[1]
    val y2 = vector2[3] - vector2[1]

    return x1 * x2 + y1 * y2
}

/**
 * @return центроид (среднюю точку для произвольного многоугольника)
 * для переданных координат вершин многоугольника.
 * */
fun getCentroid(points: FloatArray): FloatArray {
    require(points.size.rem(2) == 0, { "Array must be contain pairs of coordinates of points" })

    var centroidX = 0f
    var centroidY = 0f

    for (i in 0..points.size - 1 step 2) {
        centroidX += points[i]
        centroidY += points[i + 1]
    }

    val pointsCount = points.size / 2
    return floatArrayOf(centroidX / pointsCount, centroidY / pointsCount)
}

/**
 * @return среднее расстояние до центроида для переданных координат вершин многоугольника.
 * */
fun getAverageDistanceFromPointsToCentroid(points: FloatArray): Float {
    require(points.size >= 2, { "Require at least 1 point for caclulate distance" })
    require(points.size.rem(2) == 0, { "Array must be contain pairs of coordinates of points" })

    val centroid = getCentroid(points)

    var distance = 0f

    for (i in 0..points.size - 1 step 2) {
        distance += getDistanceBetweenTwoPoints(
                floatArrayOf(
                        points[i],
                        points[i + 1]),
                floatArrayOf(
                        centroid[0],
                        centroid[1])
        )
    }

    return distance / (points.size / 2)
}

/**
 * @return средний угол, прилегающий к центроиду для треугольников,
 * образованных между центроидом и точками по смежным индексам в @param beginPoints и @param endPoints.
 *
 * Кроме того, если вектор, образованный точккми
 * по смежным индексам в @param beginPoints и @param endPoints,
 * конец которого лежит на окружности с радиусом равным расстоянию от точки в @param beginPoints до центроида,
 * будет обращен по часовой стрелке, значение будет положительны. В противном случае, отрицательным.
 * */
fun getAverageAngleBetweenPointsPairsAndCentroid(beginPoints: FloatArray, endPoints: FloatArray): Float {
    require(beginPoints.size == endPoints.size, { "Arrays must be have equal length" })
    require(beginPoints.size.rem(2) == 0, { "Arrays must be contain pairs of coordinates of points" })
    require(endPoints.size.rem(2) == 0, { "Arrays must be contain pairs of coordinates of points" })

    val centroid = getCentroid(beginPoints)

    var angleSum = 0f

    for (i in 0..beginPoints.size - 1 step 2) {

        val scalarProduct = getScalarProduct(
                floatArrayOf(
                        centroid[0],
                        centroid[1],
                        beginPoints[i],
                        beginPoints[i + 1]),
                floatArrayOf(
                        centroid[0],
                        centroid[1],
                        endPoints[i],
                        endPoints[i + 1]))

        val vector1Length = getDistanceBetweenTwoPoints(
                floatArrayOf(centroid[0],
                        centroid[1]),
                floatArrayOf(beginPoints[i],
                        beginPoints[i + 1])
        )

        val vector2Length = getDistanceBetweenTwoPoints(
                floatArrayOf(
                        centroid[0],
                        centroid[1]),
                floatArrayOf(
                        endPoints[i],
                        endPoints[i + 1])
        )

        val cos = scalarProduct / (vector1Length * vector2Length)
        val angle = Math.toDegrees(Math.acos(cos.toDouble())).toFloat()

        val signedAngle = angle * getTriangleDirection(floatArrayOf(
                beginPoints[i],
                beginPoints[i + 1],
                endPoints[i],
                endPoints[i + 1],
                centroid[0],
                centroid[1]))


        angleSum += signedAngle
    }

    return angleSum / (beginPoints.size / 2)
}

/**
 * @return направление переданного треугольника.
 * В случае, если вершины переданы по часовой стрелке, возвращает 1, если против - -1.
 *
 * Использует для подсчета нахождение ориентированной площади треугольника.
 * */
private fun getTriangleDirection(array: FloatArray): Float {
    require(array.size == 6, { "Array must be contain 3 points" })

    val a = array[0] - array[4]
    val b = array[2] - array[4]
    val c = array[1] - array[5]
    val d = array[3] - array[5]

    return Math.copySign(1f, a * d - b * c)
}